### PR TITLE
Improve JavaDoc of TaskInputs.property

### PR DESCRIPTION
**Pull request by britter**

### CODE REVIEW

File: subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputs.java [hunk 1]
```diff
@@ -84,11 +84,12 @@
...
-     * <p>The given value for the property must be Serializable, so that it can be persisted. It should also provide a
-     * useful {@code equals()} method.</p>
+     * <p>The given value must be a simple value, like a String or Integer, or serializable. For complex values,
+     *  Gradle compares the serialized forms for detecting changes and the {@code equals()} method is ignored.
      *
-     * <p>You can specify a closure or {@code Callable} as the value of the property. In which case, the closure or
-     * {@code Callable} is executed to determine the actual property value.</p>
+     * <p>You can specify a {@link org.gradle.api.provider.Provider}, {@link java.util.concurrent.Callable}, Groovy closure,
+     * or Kotlin function as the value of the property. In which case, the function is executed to determine the actual
+     * property value.</p>
```
Comment (wolfs) [inline subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputs.java:92]: Maybe we should split those things in two parts. How about
```suggestion
     * <p>If the value is not known when registering the input, you should register a {@link org.gradle.api.provider.Provider} which Gradle evaluates lazily.
     * You can specify a {@link java.util.concurrent.Callable}, a Groovy closure,
     * or a Kotlin function as the value of the property. In which case, the function is executed to determine the actual
     * property value. You should prefer using a {@link org.gradle.api.provider.Provider}.</p>
```

Comment (wolfs) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputs.java:92]: Sounds good to me as well.

Comment (britter) [reply subprojects/core-api/src/main/java/org/gradle/api/tasks/TaskInputs.java:92]: How about:
```suggestion
     * <p>If the value is not known when registering the input, a {@link org.gradle.api.provider.Provider} can be
     * passed instead. Gradle will then resolve the provider at the latest possible time in order to determine the actual
     * property value.</p>
```

**Review by wolfs — APPROVED** [review]
LGTM!

**Comment by britter** [comment]
@bot-gradle test and merge

**Pull request merged**